### PR TITLE
Add status string to State field for inspect

### PR DESCRIPTION
**Pull request by hqhq**
Fixes: #13579

Signed-off-by: Qiang Huang h.huangqiang@huawei.com

### CODE REVIEW

File: docs/reference/api/docker_remote_api_v1.20.md [hunk 1]
```diff
@@ -425,8 +425,9 @@ Return low-level information on the container `id`
 			"Paused": false,
 			"Pid": 0,
 			"Restarting": false,
-			"Running": false,
-			"StartedAt": "2015-01-06T15:47:32.072697474Z"
+			"Running": true,
+			"StartedAt": "2015-01-06T15:47:32.072697474Z",
+			"Status": "running"
```
Comment (thaJeztah) [inline docs/reference/api/docker_remote_api_v1.20.md]: Please note that these changes should go into `docker_remote_api_v1.21.md`, once https://github.com/docker/docker/pull/14981 is merged

Comment (hqhq) [inline docs/reference/api/docker_remote_api_v1.20.md]: Of course, I'll rebase once #14981 is merged, thanks for that info.

**Comment by duglin** [comment]
At some point in the future I think it would be good to consider deprecating all of the bool's but probably not right now.

LGTM

File: docs/reference/api/docker_remote_api_v1.20.md (hunk 1 above)
Comment (thaJeztah) [inline docs/reference/api/docker_remote_api_v1.20.md]: @hqhq https://github.com/docker/docker/pull/14981 was merged (just a heads up, no hurry :-))

**Comment by tiborvass** [comment]
design +1, but would love to have another +1 

@vieux @aluzzardi  what do you think of this?

**Comment by thaJeztah** [comment]
fwiw, I'm +1 for adding the string for state; much friendlier to work with (the booleans easily lead to confusion.)

**Comment by aluzzardi** [comment]
I like this - not sure about the name `Status` though. This is more like a pretty print.

@duglin I think this should only be an addition to the bools, not a replacement, even long term. They are machine readable while the status string is not (e.g. `Up (2 hours)`).

ping @vieux

**Comment by tiborvass** [comment]
@aluzzardi i think you're misunderstanding this PR. The `status` field is not supposed to be `up (2 hours)`, but `running`. So indeed, it would have been preferrable to not have the booleans in the first place. If you want to check whether a container is running you can check `r.Status == "running"`.

**Comment by duglin** [comment]
@aluzzardi I agree with @tiborvass, StateString(), which is used to populate this new field, returned a well defined, machine readable, string.

**Comment by aluzzardi** [comment]
Gotcha. Should we export the other status as well (`Up (2 hours)`)?

Right now there are some inconsistencies - `ps` will give you that information but **not** `inspect` (and the other way around - `ps` doesn't expose `ContainerState`).

Perhaps we could consolidate the two and export `ContainerState` for `ps` as well as `inspect`

**Comment by tiborvass** [comment]
@aluzzardi I'm game, IF we remove all those booleans :) Also I'm not sure we should expose that `Up (2 hours)` string in the API. We should expose the start datetime and have the client construct the string (in a perfect world).

**Comment by thaJeztah** [comment]
> not sure we should expose that "Up (2 hours)" string in the API. We should expose the start datetime and have the client construct the string (in a perfect world)

Agree with @tiborvass here, I think that's not something the API should return. Better export a package that people can use to construct these strings if they're developing on top of the API and want to use the same formatting that Docker uses?

**Comment by icecrime** [comment]
@aluzzardi Can you please clarify in a comment what you'd like to see done so that @hqhq can update accordingly?

**Comment by aluzzardi** [comment]
Not sure if we want to do this now, but basically we have **2** ways of representing containers:

`GET /containers/json` which is used by `docker ps` and  `GET /containers/{name:.*}`, which is used by `inspect`.

`GET /containers/json` returns the following `Container` object:
```json
[
  {
    "Id": "8c8a7a0ee4650638c4d9cad409c8ad444c286f40df9daa57e06bb2bbbb957c08",
    "Names": [
      "\/sharp_meitner"
    ],
    "Image": "busybox",
    "Command": "\/bin\/sh",
    "Created": 1438995182,
    "Ports": [
      
    ],
    "Labels": {
      "com.docker.swarm.id": "1909dd69075e7582e6083e248c7b68144babd0716da6d26cc8f11fe78245e55b"
    },
    "Status": "Up 1 seconds"
  }
]
```

While `/containers/<id>/json` returns this:
```json
{
  "Id": "8c8a7a0ee4650638c4d9cad409c8ad444c286f40df9daa57e06bb2bbbb957c08",
  "Created": "2015-08-08T00:53:02.407047092Z",
  "Path": "\/bin\/sh",
  "Args": [
    
  ],
  "State": {
    "Running": true,
    [...]
  },
  "Image": "8c2e06607696bd4afb3d03b687e361cc43cf8ec1a4a725bc96e39f05ba97dd55",
  "NetworkSettings": {
    [...]
  },
  [...]
  "Name": "\/sharp_meitner",
  [...]
  "HostConfig": {
   [...]
  },
  "Config": {
    "Hostname": "8c8a7a0ee465",
    [...]
    "Labels": {
      "com.docker.swarm.id": "1909dd69075e7582e6083e248c7b68144babd0716da6d26cc8f11fe78245e55b"
    }
  }
}
```

Not only `ps` only returns **partial** data (which I could probably understand), but they way it returns data is **completely** different. Some examples on top of my head:

- Relevant to this PR: `ps` returns ` "Status": "Up 1 seconds"` while `inspect` returns `"State": {"Running": true, ...}`
- Created is an integer in `ps` while it's a string in `inspect`
- `ps` return names in a string array named `Names` while inspect has a string field named `Name`
- `ps` returns a `Command` field containing the full command-line while `inspect` has both a `Path` (string) and `Args` (string array) field

And so on and so on ...

My original comment was meant to merge `Status` in ps and `State` in inspect into a single object, but given all those problems, we might as well do nothing and just re-design this API.

As for the trivia on *how did things get so wrong?*, I believe `ps` was the only official API to get a list of containers while `inspect` was meant to be a **temporary** command to dump the internal data structure of a container in order to debug in the early days. I guess it's not so temporary anymore :)

**Comment by duglin** [comment]
As part of a different set of PRs, it might be good
- to make `GET /containers/json` and `GET /containers/{name:.*}` return the same data for each container.  If things are missing from the `inspect` variant, like "up 1 seconds" then we should add a "stateChanged" timestamp to both and let the client calculate the text string to display.
- get rid of the booleans from the internal structure. We can keep utility funcs if necessary to simulate their presence but I don't see a lot of value in keeping them.

but for this PR, LGTM

**Comment by cpuguy83** [comment]
A little confused.. are we good with adding this string?
LGTM

**Comment by calavera** [comment]
LGTM

**Comment by LK4D4** [comment]
LGTM

**Pull request merged**